### PR TITLE
#13 server side export

### PR DESCRIPTION
**Pull request by cdaniel**

### CODE REVIEW

**Comment by cdaniel** [comment]
This is work in progress - I still have not figured out why created svg files miss the styling.

**Comment by cdaniel** [comment]
Ok, the generated svg file looks good in inkscape. I guess it is the poor state of the node/jsdom/d3 on the server :-(

File: server/export.js [hunk 1]
```diff
@@ -255,82 +263,176 @@ var draw = function(res, filename, map, x, y){
...
+			var y = 1024;
+			var thumbnailMargin = 20;
+			
+			var el = window.document.querySelector('#map-container');
+			var svgimg = d3.select(el).append('svg:svg').attr('xmlns', 'http://www.w3.org/2000/svg').attr('class', 'map');
+			
+			
+			var defs = svgimg.append('defs');
+			
+			var css = defs.append('style').attr('type','txt/css');
```
Comment (ScottWeinstein) [inline server/export.js]: the `<style>` element should be a child of the `<svg>` element, not `<defs>`, then the css will work

**Comment by cdaniel** [comment]
Still not perfect, but I want to unblock you Scott. Canvas is out, and I hope for good!

File: client/favicon.svg [hunk 2]
```diff
@@ -1,17 +0,0 @@
-<?xml version="1.0"?>
```
Comment (ScottWeinstein) [inline client/favicon.svg]: usefull for empty maps to have an image

File: client/mapeditor.jade [hunk 3]
```diff
@@ -58,15 +58,8 @@ html
...
-                            a.dropdown-toggle(data-toggle="dropdown" href="#") Export <b class="caret"></b>
-                            ul.dropdown-menu(role="menu")
-                                li
-                                    a#exportsmall(download="#{map.history[0].name}.png" href="/api/map/#{map._id}/export/1/#{map._id}.png") Small
-                                li
-                                    a#exportmedium(download="#map.history[0].name}.png" href="/api/map/#{map._id}/export/2/#{map._id}.png") Medium
-                                li
-                                    a#exportsmall(download="#{map.history[0].name}.png" href="/api/map/#{map._id}/export/3/#{map._id}.png") Large
+                        li
+                            a#download(download="#{map.history[0].name}.png" href="/api/svg/#{map._id}/#{map._id}.svg") Download
```
Comment (ScottWeinstein) [inline client/mapeditor.jade]: should be .svg filename
also, may wish to consider two link options here, a download and a plain link option

File: package.json [hunk 4]
```diff
@@ -31,11 +31,14 @@
     "log4js": "*",
     "jade" : "*",
     "request" : "*",
-    "mailchimp-api" : "*"
+    "mailchimp-api" : "*",
+    "d3" : "*",
```
Comment (ScottWeinstein) [inline package.json]: should specify the version so that accidental breaks are not introduced

File: server/export.js [hunk 5]
```diff
@@ -14,345 +14,199 @@ limitations under the License.*/
...
+				};
 			}
+				
+			
+			var x = 1280;
+			var y = 1024;
+			var thumbnailMargin = 20;
+			
+			var el = window.document.querySelector('#map-container');
+			var svgimg = d3.select(el).append('svg:svg').attr('xmlns', 'http://www.w3.org/2000/svg').attr('class', 'map');
```
Comment (ScottWeinstein) [inline server/export.js]: what's your thinking on css vs in-line styles? Advantages of css mean less JS code and smaller svg file for larger maps

File: server/export.js [hunk 6]
```diff
@@ -14,345 +14,199 @@ limitations under the License.*/
...
+// TODO: 'Created at www.wardleymaps.com'
+// TODO: map name
+// TODO: map description
+// TODO: scope:actions, scope:jsPlumb_DefaultScope
+
+
+
+var draw = function(res, filename, map){
+	var htmlStub = '<html><head><script></script></head><body><div id="map-container"></div></body></html>';
+	var scripts = ['//cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.js', '//cdnjs.cloudflare.com/ajax/libs/lodash.js/3.2.0/lodash.min.js'];
```
Comment (ScottWeinstein) [inline server/export.js]: this isn't needed

File: client/mapeditor.jade (hunk 3 above)
Comment (cdaniel) [inline client/mapeditor.jade]: good catch, I will correct this today evening.

File: server/export.js (hunk 5 above)
Comment (cdaniel) [inline server/export.js]: the only reason why there are in-line styles is that I could not get internal css working. I've tried your advice of embedding style directly under svg, but the result is the same - black nodes and nothing else displayed. Inkscape rendered the map correctly, but browser was black-nodes-only. I gave up on this, especially as nobody complained about the map size ;-).

Comment (ScottWeinstein) [inline server/export.js]: What browser are you using?

**Comment by cdaniel** [comment]
firefox-35.0.1-5.fc21.x86_64
google-chrome-stable-40.0.2214.115-1.x86_64


svgcomparison

This image is actually svg, but i've changed it to png because github does not accept svgs. You can try by yourself.
thumbnail

**Comment by ScottWeinstein** [comment]
I took a look at this. If you change this line from

`<style type="txt/css">`
to
`<style type="text/css">`

and move it above the defs, like so

```
<svg xmlns="http://www.w3.org/2000/svg" class="map" width="1280" height="1024">
    <style type="text/css">
        .map circle {...}
    </style>
    <defs>
        <marker id="arrow" orient="auto" markerWidth="6" markerHeight="8" refX="0.1" refY="2"><path d="M0,0 V4 L2,2 Z" fill="grey"/></marker>
    </defs>
```

then it renders well. In FF,Chrome, and IE

**Comment by cdaniel** [comment]
You're awesome!

https://github.com/cdaniel/wardleymapstool/commit/e0579b24b8586f456bcd80d36b215d1a91daec4b

**Pull request merged**